superduperdb/ext/llm/base.py

```diff
@@ -0,0 +1,317 @@
+
+if typing.TYPE_CHECKING:
+    from superduperdb.base.datalayer import Datalayer
+
+# Disable httpx info level logging
+getLogger("httpx").setLevel(WARNING)
+
+
+def ensure_initialized(func):
+    """Decorator to ensure that the model is initialized before calling the function"""
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if not hasattr(self, "_is_initialized") or not self._is_initialized:
+            model_message = f"{self.__class__.__name__} : {self.identifier}"
+            logging.info(f"Initializing {model_message}")
+            self.init()
+            self._is_initialized = True
+            logging.info(f"Initialized  {model_message} successfully")
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+@dc.dataclass
+class _BaseLLM(Component, _Predictor, metaclass=abc.ABCMeta):
+    """
+    :param prompt_template: The template to use for the prompt.
+    :param prompt_func: The function to use for the prompt.
+    :param max_batch_size: The maximum batch size to use for batch generation.
+    :param inference_kwargs: Parameters used during inference.
+    """
+
+    prompt_template: str = "{input}"
+    prompt_func: Optional[Callable] = dc.field(default=None)
+    max_batch_size: Optional[int] = 64
+    inference_kwargs: dict = dc.field(default_factory=dict)
+
+    def __post_init__(self):
+        super().__post_init__()
+        self.takes_context = True
+        self.identifier = self.identifier.replace("/", "-")
+        assert "{input}" in self.prompt_template, "Template must contain {input}"
+
+    def to_call(self, X, *args, **kwargs):
+        raise NotImplementedError
+
+    def post_create(self, db: "Datalayer") -> None:
+        # TODO: Do not make sense to add this logic here,
+        # Need a auto DataType to handle this
+        from superduperdb.backends.ibis.data_backend import IbisDataBackend
+        from superduperdb.backends.ibis.field_types import dtype
+
+        if isinstance(db.databackend, IbisDataBackend) and self.encoder is None:
+            self.encoder = dtype("str")
+
+        # since then the `.add` clause is not necessary
+        output_component = db.databackend.create_model_table_or_collection(
+            self  # type: ignore[arg-type]
+        )
+
+        if output_component is not None:
+            db.add(output_component)
+
+    @abc.abstractmethod
+    def init(self):
+        ...
+
+    @abc.abstractmethod
+    def _generate(self, prompt: str, **kwargs: Any) -> str:
+        ...
+
+    def _batch_generate(self, prompts: List[str], **kwargs: Any) -> List[str]:
+        """
+        Base method to batch generate text from a list of prompts.
+        If the model can run batch generation efficiently, pls override this method.
+        """
+        return [self._generate(prompt, **kwargs) for prompt in prompts]
+
+    @ensure_initialized
+    def _predict(self, X: Union[str, List[str]], one: bool = False, **kwargs: Any):
```

Comment:
A point for later: none of the LLM implementations we have seem to support `postprocess` or `preprocess`.

Reply:
I thought it was compatible with this before, but the call chain was too deep. But you are right, I will change the implementation slightly to support them